src/main/java/com/segment/analytics/android/integrations/optimizelyx/OptimizelyXIntegration.java

```diff
@@ -1,63 +1,226 @@
+ *     Android SDK</a>
  */
 public class OptimizelyXIntegration extends Integration<Void> {
-  private static final String OPTIMIZELY_KEY = "Optimizely";
 
-  private final Context context;
+  private static final String OPTIMIZELYX_KEY = "Optimizely X";
+  NotificationListener listener;
+  private OptimizelyClient client;
   private final Logger logger;
-  private final Analytics analytics;
-
-  // Optimizely must be initialized immediately on launch. So we require the token when creating the
-  // factory, and initialize it in this method.
-//  public static Factory createFactory(final Application application, String token) {
-////    Optimizely.startOptimizelyWithAPIToken(token, application);
-////    return new Factory() {
-////      @Override public Integration<?> create(ValueMap settings, Analytics analytics) {
-////        boolean listen = settings.getBoolean("listen", false);
-////        Logger logger = analytics.logger(OPTIMIZELY_KEY);
-////
-////        OptimizelyIntegration integration =
-////            new OptimizelyIntegration(analytics, application, logger);
-////        if (listen) {
-////          Optimizely.addOptimizelyEventListener(integration);
-////        }
-////        return integration;
-////      }
-//
-////      @Override public String key() {
-////      return OPTIMIZELY_KEY;
-//////      }
-//    };
-//  }
-
-  public OptimizelyXIntegration(Analytics analytics, Context context, Logger logger) {
-    this.analytics = analytics;
-    this.context = context;
+  boolean isClientValid = false;
+  boolean trackKnownUsers;
+  static boolean nonInteraction;
+  boolean listen;
+  static final Options options = new Options().setIntegration(OPTIMIZELYX_KEY, false);
+  private Map<String, String> attributes = new HashMap<>();
+  private List<TrackPayload> trackEvents = new ArrayList<>();
+  private final Handler handler = new Handler();
+
+  public static Factory factory(OptimizelyManager manager) {
+    return new Factory(manager);
+  }
+
+  private static class Factory implements Integration.Factory {
+    private final OptimizelyManager manager;
+
+    Factory(OptimizelyManager manager) {
+      this.manager = manager;
+    }
+
+    @Override
+    public Integration<?> create(ValueMap settings, Analytics analytics) {
+      Logger logger = analytics.logger(OPTIMIZELYX_KEY);
+      return new OptimizelyXIntegration(analytics, manager, settings, logger);
+    }
+
+    @Override
+    public String key() {
+      return OPTIMIZELYX_KEY;
+    }
+  }
+
+  public OptimizelyXIntegration(
+      final Analytics analytics, OptimizelyManager manager, ValueMap settings, Logger logger) {
+    this.client = manager.getOptimizely();
     this.logger = logger;
+
+    trackKnownUsers = settings.getBoolean("trackKnownUsers", false);
+    nonInteraction = settings.getBoolean("nonInteraction", true);
+    listen = settings.getBoolean("listen", true);
+
+    if (client.isValid()) {
+      isClientValid = true;
```

Comment:
Instead of using a separate variable, why not just use the client object with client.isValid()?

Reply:
This was intentional for efficiency, and to ensure we're only checking the client object directly on our polling thread. Let me know if this is not the right approach!